tool/tsh/common/putty_config_windows.go

```diff
@@ -0,0 +1,365 @@
+		ProxyTelnetCommand:  commandToRun,
+		PublicKeyFile:       ppkFilePath,
+		DetachedCertificate: certificateFilePath,
+		UserName:            login,
+	}
+
+	// now check for and create the individual session key
+	pk, err := registry.GetOrCreateRegistryKey(registryKey)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer pk.Close()
+
+	// write dwords
+	if err := registry.WriteDword(pk, "Present", sessionDwords.Present); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteDword(pk, "PortNumber", fmt.Sprintf("%v", sessionDwords.PortNumber)); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteDword(pk, "ProxyPort", fmt.Sprintf("%v", sessionDwords.ProxyPort)); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteDword(pk, "ProxyMethod", sessionDwords.ProxyMethod); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteDword(pk, "ProxyLogToTerm", sessionDwords.ProxyLogToTerm); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// write strings
+	if err := registry.WriteString(pk, "Hostname", sessionStrings.Hostname); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "Protocol", sessionStrings.Protocol); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "ProxyHost", sessionStrings.ProxyHost); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "ProxyUsername", sessionStrings.ProxyUsername); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "ProxyTelnetCommand", sessionStrings.ProxyTelnetCommand); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "PublicKeyFile", sessionStrings.PublicKeyFile); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "DetachedCertificate", sessionStrings.DetachedCertificate); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := registry.WriteString(pk, "UserName", sessionStrings.UserName); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// addHostCAPublicKey adds a host CA to the registry with a set of space-separated hostnames
+func addHostCAPublicKey(keyName string, publicKey string, hostname string) error {
+	registryKeyName := fmt.Sprintf(`%v\%v`, puttyRegistrySSHHostCAsKey, keyName)
+
+	// get the subkey with the host CA key name
+	registryKey, err := registry.GetOrCreateRegistryKey(registryKeyName)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	hostList, _, err := registryKey.GetStringsValue("MatchHosts")
+	if err != nil {
+		// ERROR_FILE_NOT_FOUND is an acceptable error, meaning that the value does not already
+		// exist and it must be created
+		if err != syscall.ERROR_FILE_NOT_FOUND {
+			log.Debugf("Can't get registry value %v: %T", registryKeyName, err)
+			return trace.Wrap(err)
+		}
+	}
+	// initialize an empty hostlist if there isn't one stored under the registry key
+	if len(hostList) == 0 {
+		hostList = []string{}
+	}
+	defer registryKey.Close()
```

Comment:
This should happen immediately following the error check on line 179, right?

If you get ERROR_FILE_NOT_FOUND, you'll return early and the handle will never be closed.

Reply:
I have a feeling I did this before and it caused some weird issue where the key was closed too early. I agree that it has the potential to leak, so I'll investigate further.